### PR TITLE
Fix ConstructorConstructor creating mismatching Collection and Map instances

### LINKED ISSUE #419: Gson can't construct subclasses of HashMap without a default constructor
```
If we have a subclass of HashMap without a default constructor, Gson uses 
LinkedHashMap as the object type. This isn't correct since we should have 
instead tried to use UnsafeAllocator.
```

Original issue reported on code.google.com by `inder123` on 11 Mar 2012 at 9:42

### LINKED ISSUE #1708: ConstructorConstructor.newDefaultImplementationConstructor(Type, Class<? super T>) breaks symmetry
The method `com.google.gson.internal.ConstructorConstructor.newDefaultImplementationConstructor(Type, Class<? super T>)` breaks symmetry.

Normally it is desired that if you can serialize `obj` as `A`, then you should be able to deserialize it as `A` again. This is however not possible if `A` extends / implements one of the types checked in `newDefaultImplementationConstructor` and does not have a no-arg constructor.
Example:
```
static class MyList extends ArrayList<Integer>{
  // Any constructor with parameters to prevent default constructor
  public MyList(int i) {}
}
public void testCustomList() {
  Gson gson = new Gson();
  MyList list = new MyList(1);
  list.add(1);
  String json = gson.toJson(list);
  // java.lang.ClassCastException: Cannot cast java.util.ArrayList to MyList
  gson.fromJson(json, MyList.class);
}
```

The proper fix would be to have `newDefaultImplementationConstructor` only create instances if the requested type is the type it creates (e.g. it creates `ArrayList` and `ArrayList` is requested) or if the requested type is a supertype of the created one (e.g. it creates `ArrayList` and `List` is requested).
Similarly `CollectionTypeAdapterFactory` would have to be adjusted to only support these types and types for which a constructor has been registered or types which have a default (= no-arg) constructor.

This would in theory not cause any issues if people already serialized **and** deserialized their objects and therefore the types they are using are matching the above stated requirements.

However, there might be cases where people only serialize objects (such as `MyList` in the example above), which would break with such a change.

Would it make sense to instead introduce an option to `GsonBuilder` to prevent serialization which is not symmetric? So people can catch such issues early?

### LINKED ISSUE #2067: `EnumMap` deserialization fails with `ClassCastException`
# Gson version
e2e851c9bc692cec68ba7b0cbb002f82b4a229e4

# Java / Android version
openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment Temurin-11.0.13+8 (build 11.0.13+8)
OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

# Description
`EnumMap` deserialization fails with `ClassCastException`:
```
java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class java.util.EnumMap
```

The underlying issue is related to #1708, it appears special handling for `EnumMap` is missing.
(I am a bit surprised that this issue has not been mentioned here anywhere before.)

# Reproduction steps
Test case:
```java
private static enum MyEnum {
  VALUE1, VALUE2
}

public void testEnumMap() throws Exception {
  EnumMap<MyEnum, String> map = new EnumMap<MyEnum, String>(MyEnum.class);
  map.put(MyEnum.VALUE1, "test");
  String json = gson.toJson(map);
  assertEquals("{\"VALUE1\":\"test\"}", json);

  Type type = new TypeToken<EnumMap<MyEnum, String>>() {}.getType();
  EnumMap<?, ?> actualMap = gson.fromJson("{\"VALUE1\":\"test\"}", type);
  Map<?, ?> expectedMap = Collections.singletonMap(MyEnum.VALUE1, "test");
  assertEquals(expectedMap, actualMap);
}
```

# Exception stack trace
Not really useful because `ClassCastException` occurs in user code.

### LINKED ISSUE #2730: Gson Conversion Issue: Unable to Convert java.util.ArrayList to RealmList Implicitly During Deserialization in Android Studio
# Gson version
2.10.1 AND 2.11.0

# Java / Android version
Android version 13

# Used tools

- [ ] Gradle; version: 7.5
- [ ] Realm-Kotlin; version: 1.16.0

# Stack Trace
```
java.lang.IllegalArgumentException: field com.nextgenpos.nextgenserverpos.data.model.order.nestedOrderModels.Item.customisedAddOns has type io.realm.kotlin.types.RealmList, got java.util.ArrayList
2024-08-29 15:58:53.232 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at java.lang.reflect.Field.set(Native Method)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$2.readIntoField(ReflectiveTypeAdapterFactory.java:278)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:558)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:516)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:40)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:83)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:59)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$2.readIntoField(ReflectiveTypeAdapterFactory.java:267)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:558)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:516)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.Gson.fromJson(Gson.java:1361)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.Gson.fromJson(Gson.java:1262)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.Gson.fromJson(Gson.java:1171)
2024-08-29 15:58:53.233 27746-28472 System.err              com.nextgenpos.nextgenserverpos      W  	at com.google.gson.Gson.fromJson(Gson.java:1107)
```

# Description
When deserialising a JSON into an object containing RealmList Gson is unable to populate it correctly and throws the following error. In fact the same error occurs in case of RealmDictionary too.

## Expected behavior
The deserialised JSON should correctly populate the any list type to RealmList

## Actual behaviour
IllegalArgumentException is thrown

### PR DESCRIPTION
**Pull request by Marcono1234**
Fixes #419
Fixes #1708
Fixes #2730 (the proper exception that interfaces cannot be instantiated will be thrown instead)

Gson's `ProtoTypeAdapter` (in the non-published `proto` module) indirectly relied on `ConstructorConstructor`'s broken behavior:
During deserialization of 'repeated' fields it [peeks at the type of the internal Java field](https://github.com/google/gson/blob/6af67052e5e088701d65addd0d40ef7498adb5ac/proto/src/main/java/com/google/gson/protobuf/ProtoTypeAdapter.java#L333-L334). The problem is that this field does not have the type `List<E>` is all cases but for some types has specialized Protobuf subinterfaces. For example for `List<Long>` the internal field has the type `com.google.protobuf.Internal.LongList`.
So previously Gson erroneously created `ArrayList` instances even though Protobuf's `LongList` is not related to `ArrayList`. This worked because `LongList` is only an implementation detail, the Protobuf builder API seems to expect just `List<E>` so it did not cause any problems that Gson used an `ArrayList`. (Still the behavior that Gson created `ArrayList` when asked for a `LongList` is arguably wrong.)

### CODE REVIEW

**Comment by eamonnmcmanus** [comment]
The current behaviour seems hard to understand, but I am concerned if we change it to be more logical we may break projects that are implicitly depending on the way it is now. I ran this PR against Google's internal tests and encountered four failures in four unrelated projects. All appeared to be related to protobufs, some directly via the `ProtoTypeAdapter` in this project and others via what appear to be modified clones of that class. My inclination is to leave this alone.

**Comment by Marcono1234** [comment]
Hmm, ok that is unfortunate. I have split `EnumMap` support into a separate pull request (#2071). If you want you can close this pull request here then, if you don't think there is any chance this will be integrated.

**Comment by Marcono1234** [comment]
@eamonnmcmanus, can / should I revive this pull request? As seen with #2730, the current behavior is broken and leads to quite confusing exceptions, for example in cases where you deserialize a custom `List` subtype but Gson creates an `ArrayList` instead.
Only in rare cases this actually works, where the user code is not completely type-safe.

In case this still applies to the Protobuf handling, then if necessary we could hardcode support for that.

**Comment by eamonnmcmanus** [comment]
OK, if you want to bring this up to date I will retry the experiment of running it against Google's tests and we can possibly come up with an adjustment that makes those tests pass.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 1]
```diff
@@ -1360,7 +1358,19 @@ public <T> T fromJson(JsonReader reader, TypeToken<T> typeOfT)
       JsonToken unused = reader.peek();
       isEmpty = false;
       TypeAdapter<T> typeAdapter = getAdapter(typeOfT);
-      return typeAdapter.read(reader);
+      T object = typeAdapter.read(reader);
+      Class<?> expectedTypeWrapped = Primitives.wrap(typeOfT.getRawType());
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:1362]: It seems all other `fromJson` methods delegate to this one, so I have added the type check (which was previously `Primitives.wrap(classOfT).cast(object)`) here instead.
However, this method did not actually have this check before.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 2]
```diff
@@ -1360,7 +1358,19 @@ public <T> T fromJson(JsonReader reader, TypeToken<T> typeOfT)
...
+      if (object != null && !expectedTypeWrapped.isInstance(object)) {
+        throw new ClassCastException(
+            "Type adapter '"
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:1365]: Not completely sure if this custom `ClassCastException` is really worth it, maybe a `Class#cast` call would suffice, as done before.

This check will only help when directly using `fromJson` for an incorrect adapter, but if the adapter is used indirectly (for example for a list value) it would still cause other less descriptive exceptions later.

File: gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java [hunk 3]
```diff
@@ -334,78 +327,136 @@ private static <T> ObjectConstructor<T> newDefaultImplementationConstructor(
...
-          }
-        };
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newMapConstructor(type, rawType);
+      return constructor;
+    }
+
+    // Unsupported type; try other means of creating constructor
+    return null;
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:342]: In https://github.com/google/gson/issues/419#issuecomment-83705342 the concern was mentioned that the fallback will be `Unsafe` which might leave the Collection or Map implementation in a broken state because fields have not been initialized. And the built-in Collection and Map adapters might then encounter confusing exceptions when trying to add elements.

Not sure if we should throw an exception then in case of Collection or Map? Or we could maybe add an `allowUnsafe` parameter here (which is considered in addition to `GsonBuilder#disableJdkUnsafe`), which the built-in Collection and Map adapters can set to `false` because calling `add` and `put` will likely fail afterwards.

File: gson/src/test/java/com/google/gson/functional/MapTest.java [hunk 4]
```diff
@@ -210,6 +213,49 @@ public void testMapDeserializationWithUnquotedLongKeys() {
     assertThat(map.get(longKey)).isEqualTo("456");
   }
 
+  @Test
+  public void testMapStringKeyDeserialization() {
```
Comment (Marcono1234) [inline gson/src/test/java/com/google/gson/functional/MapTest.java:217]: There is some overlap between the newly added tests here and in `ConstructorConstructorTest`, should we keep both nonetheless?

**Review by Marcono1234 — COMMENTED** [review]
Have updated this pull request now, and also tried to solve the Protobuf issue in a (hopefully) reasonable way, see changes in `ProtoTypeAdapter`.

However, it will require that users update their local copy of `ProtoTypeAdapter`.
Other ways of solving this would probably require special hardcoded Protobuf behavior in `ConstructorConstructor`, which is possible, but probably not a very clean approach.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 5]
```diff
@@ -1360,7 +1358,19 @@ public <T> T fromJson(JsonReader reader, TypeToken<T> typeOfT)
...
+                + typeAdapter
+                + "' returned wrong type; requested "
+                + typeOfT.getRawType()
+                + " but got instance of "
+                + object.getClass()
+                + "\nVerify that the adapter was registed for the correct type.");
```
Comment (eamonnmcmanus) [inline gson/src/main/java/com/google/gson/Gson.java:1371]: Typo: registered

**Review by eamonnmcmanus — COMMENTED** [review]
I ran this against all of Google's internal tests and there were several failures. I plan to investigate in more detail next week.

File: gson/src/main/java/com/google/gson/Gson.java (hunk 1 above)
Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/Gson.java:1362]: I ran this change against all of Google's internal tests and found a _lot_ of failures with this exception, or with an exception about not being able to instantiate the abstract class `com.google.common.collect.ImmutableList`. In both cases I think the code in question should be fixed, because it is relying on accidental behaviour. A pattern I saw quite often was this:

```java
@AutoValue
public abstract Foo {
  public abstract ImmutableList<Bar> bars();

  public static TypeAdapter<Foo> typeAdapter(Gson gson) {
    return new AutoValue_Foo.GsonTypeAdapter(gson);
  }

  @AutoValue.Builder
  public interface Builder {
    public Builder setBars(List<Bar> bars);
    public Foo build();
  }
}
```

This is using [auto-value-gson](https://github.com/rharter/auto-value-gson) to serialize and deserialize AutoValue classes. It works even if no `TypeAdapterFactory` has been registered for `ImmutableList`, because Gson deserializes a plain `List`. Auto-value-gson constructs the `Foo` instance using its builder, and the deserialized `List` works with the `setBars` method. That method calls `ImmutableList.copyOf(List)` so we do get the right `ImmutableList` type. But the additional checks in this PR don't let us get that far.

As I say, code like that only works accidentally. My plan is to add a `GuavaCollectionsTypeAdapterFactory` to `extras` and update the internal code to use that. I already found several custom `ImmutableListTypeAdapterFactory` and `ImmutableMapTypeAdapterFactory` classes which could be switched to use this.

Some of the other failures were just plain using the wrong types, which again happened to work through erasure and the like.

So in short, I think this change is OK, but it may require some client code to be updated. Most of the updates should be straightforward and will lead to client code that is more correct.

I think it will be a while before we can complete this because of all the adjustments that will need to be made.

**Comment by eamonnmcmanus** [comment]
I'm looking to pick this up again. @Marcono1234, would you be able to update it?

File: gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java [hunk 6]
```diff
@@ -303,33 +296,84 @@ private static <T> ObjectConstructor<T> newDefaultImplementationConstructor(
...
+  }
+
+  private static ObjectConstructor<? extends Collection<? extends Object>> newCollectionConstructor(
+      Class<?> rawType) {
+
+    // First try List implementation
+    if (rawType.isAssignableFrom(ArrayList.class)) {
+      return ArrayList::new;
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:319]: Not sure if the constructor references here are a good idea, or if it would be easier to understand if a lambda `() -> new ArrayList<>()` was used to make it clearer which of the constructors is used.

**Comment by Marcono1234** [comment]
Sorry for the delay. I have updated the pull request now.

However with the recent other changes we had to revert, I am not sure if merging these changes here is really a good idea. I still think the changes here are the "more correct" implementation, but maybe there are simply parts of Gson which cannot be improved or fixed in the near future without breaking things for users who rely on the current broken behavior.

**Comment by eamonnmcmanus** [comment]
> Sorry for the delay. I have updated the pull request now.

Thanks!

> However with the recent other changes we had to revert, I am not sure if merging these changes here is really a good idea. I still think the changes here are the "more correct" implementation, but maybe there are simply parts of Gson which cannot be improved or fixed in the near future without breaking things for users who rely on the current broken behavior.

I was a little hesitant too, especially when I saw a moderately large number of failures when running all Google's internal tests against these changes. However, I think it is quite problematic that something like `gson.fromJson(x, new TypeToken<ImmutableList<String>>() {})` tells the compiler that it is producing an `ImmutableList<String>` while today it is actually producing an `ArrayList<String>`. _Most_ of the time people will have discovered this because of a `ClassCastException`, but I found a few places where people were deliberately or accidentally dodging that. I talked more about that in my [earlier comment](https://github.com/google/gson/pull/2068#discussion_r1750716301). I might add the `GuavaCollectionsTypeAdapterFactory` to the `extras` module. Meanwhile, it is fairly easy to write your own `JsonDeserializer<ImmutableList<?>>` or copy one such as [this one](https://github.com/acebaggins/guava-gson-serializers/blob/master/src/main/java/com/baggonius/gson/immutable/ImmutableListDeserializer.java).

We'll certainly want to bump the minor version number if we merge and release this change.

File: gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java (hunk 6 above)
Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:319]: I'm fine with either.

File: gson/src/main/java/com/google/gson/Gson.java (hunk 2 above)
Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/Gson.java:1363]: FWIW this detail was quite useful to me when fixing incorrect code impacted by this change.

File: gson/src/test/java/com/google/gson/functional/MapTest.java (hunk 4 above)
Comment (eamonnmcmanus) [reply gson/src/test/java/com/google/gson/functional/MapTest.java:216]: Up to you. Also fine if you want to defer the decision.

**Review by eamonnmcmanus — APPROVED** [review]

File: gson/src/test/java/com/google/gson/functional/MapTest.java (hunk 4 above)
Comment (Marcono1234) [reply gson/src/test/java/com/google/gson/functional/MapTest.java:216]: Ok, I think I will keep this additional test then for now.

File: gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java (hunk 3 above)
Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:311]: @eamonnmcmanus, do you think this is something to consider?

Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:311]: I'm not sure I understand the concern here. If we reach this point then `rawType` is neither a `Collection` nor a `Map`. Could you explain what sort of type you might consider throwing an exception for here?

Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:333]: Seems like there is a [bug with GitHub](https://github.com/orgs/community/discussions/150115#discussioncomment-12294289) and it is not showing the comment I was referring to above; here is an [archived version](https://web.archive.org/web/20200917230215/https://github.com/google/gson/issues/419#issuecomment-83705342):

> I'm anxious about using UnsafeAllocator on map and collection types. Types like HashMap expect their constructors to run (and initialize state) but they won't necessarily get run if we rely on UnsafeAllocator. (We'd be relying on the presence of a no-args constructor).

When users define custom `Collection` or `Map` implementations without no-args constructor, then `Unsafe` would be used to create the instance. But because the Collection and Map adapter calls Collection and Map methods during deserialization, that most likely leads to confusing exceptions because Unsafe did not properly initialize the instances.

I have pushed a commit now which prevents Unsafe usage in this case; I hope that is fine. If you don't think it should be included please let me know and I revert it again.

Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:333]: OK, I think I understand the concern now. It is not about this statement particularly, but the fact that the earlier `return` statements can return null if `newCollectionConstructor` or `newMapConstructor` does, and then we'll fall back to the `Unsafe` path.

We don't see this issue with Guava's `ImmutableList` etc because the public types are abstract, so `Unsafe` obviously can't instantiate them. I generally hate the nonsense with `Unsafe` allocation so I'm all for changes that allow less of it. However I'll want to run this latest version past Google's tests again to see if there are further unforeseen consequences.

**Review by eamonnmcmanus — APPROVED** [review]
